### PR TITLE
fix ipvs_svc deletion

### PR DESCRIPTION
**Pull request by gongguan**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
In ipvs mode, when service deleted but  rs(pod) exist, related ipvs rule and address bound in ipvs0 will not be deleted.
I think it's unreasonable that kubernetes-service deleted but the service ip is available(ipvs rule not deleted). When you create same service again, new virtual service(totally two ipvs virtual service) will proxy to the same rs, too. And the deleted service ip will exist on both ipvs rule and ipvs0 device.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Fix a bug in the IPVS proxier where virtual servers are not cleaned up even though the corresponding Service object was deleted.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by gongguan** [comment]
another question: graceful termination to all real servers can be deleted too?
As service deleted, can be ipvs virtual server deleted directly, then svc addr unbound from ipvs0 net link?

**Comment by gongguan** [comment]
there is a logical error in current code:
graceful termination will never executed

File: pkg/proxy/ipvs/proxier.go [hunk 1]
```diff
@@ -1772,11 +1772,6 @@ func (proxier *Proxier) cleanLegacyService(activeServices map[string]bool, curre
 		if _, ok := activeServices[cs]; !ok {
 			rsList, _ := proxier.ipvs.GetRealServers(svc)
 
-			// If we still have real servers graceful termination is not done
```
Comment (andrewsykim) [inline pkg/proxy/ipvs/proxier.go]: In this case we would still try to delete a service even if it has real servers. This check should at least happen after the graceful termination checks below on line 1791. What do you think?

Comment (gongguan) [reply pkg/proxy/ipvs/proxier.go]: if this check moved to line 1791, deleted service ip will still bind on the kube-ipvs0 link device even rs(pod) deleted later

**Review by gongguan — COMMENTED** [review]
/retest

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: Oh, because on the next resync `currentServices` won't have the service anymore?

Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: In that case, let's just remove all the graceful termination logic for the delete case?

Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: Nevermind, `currentServices` should still have the deleted service IP cause of it queries via netlink. In what scenario would we stilll bind on kube-ipvs0 even if there are no real servers?

Comment (gongguan) [reply pkg/proxy/ipvs/proxier.go]: Actually currentServices get from ipvs virtualService list rather than query netlink. In my opinion, ip address unbind along with ipvs rules deleted. When the same service recreate, new service ip will be bound on netlink.
In this issue we can see the deleted service IP all bound on netlink.[#80407](https://github.com/kubernetes/kubernetes/issues/80407)

**Comment by lbernail** [comment]
There is something wrong with this part of the code:
- when we delete a service both the service and the endpoints will be deleted
- endpoint deletion will trigger graceful termination
- so this test `if len(rsList) > 0` makes sense: we don't want to remove the VS before all RS have gracefully terminated
- however this loop doesn't `for _, rs := range rsList`: when we reach this part of the code rsList is empty anyway

=> Unless I miss something we should just remove the useless for loop and the VS will be garbage collected once all RS have gracefully terminated (the current seems to do what we want be we can remove a few lines that are never used)

@gongguan have you actually seen a situation where deleting a service doesn't remove the VirtualServer?

**Comment by lbernail** [comment]
The address should be unbound here: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L1799

File: pkg/proxy/ipvs/proxier.go (hunk 1 above)
Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: If I remember correctly, `GetVirtualServers` gets the virtual servers from netlink though?

Comment (andrewsykim) [reply pkg/proxy/ipvs/proxier.go]: `If I remember correctly, GetVirtualServers gets the virtual servers from netlink though?`

**Comment by andrewsykim** [comment]
> Unless I miss something we should just remove the useless for loop and the VS will be garbage collected once all RS have gracefully terminated (the current seems to do what we want be we can remove a few lines that are never used)

This seems reasonable to me, but also should we even care about graceful termination if the _Service_ is deleted?

**Comment by andrewsykim** [comment]
Ran some of my own tests on this. If you delete a service directly without deleting it's real servers first then the IPVS virtual server / bind address is never deleted on the host because of the `len(rsList) > 0` check. I think for the case where a Service object is deleted, there's no reason to initiate graceful termination at that point. We should remove the `len(rsList) > 0` check AND the graceful termination in `cleanupLegacyServices`. Thoughts?

**Comment by gongguan** [comment]
> Ran some of my own tests on this. If you delete a service directly without deleting it's real servers first then the IPVS virtual server / bind address is never deleted on the host because of the `len(rsList) > 0` check. I think for the case where a Service object is deleted, there's no reason to initiate graceful termination at that point. We should remove the `len(rsList) > 0` check AND the graceful termination in `cleanupLegacyServices`. Thoughts?

This could be a solution. @lbernail 's thought is we can move `len(rsList) > 0` to line 1787. Then graceful termination will be done after service deleted. In next sync loop, the VirtualService without RS can be cleaned. I am trying to verify is. Please wait a minute.

**Comment by gongguan** [comment]
I tried to move `len(rsList) > 0` check to line 1787, the result is graceful termination will be done after service deleted, and graceful deleted VirtualService can be deleted during next sync loop's `cleanLegacyService`.
Compared with delete `len(rsList) > 0` and `graceful termination`, which way is better? @andrewsykim  @lbernail

**Comment by andrewsykim** [comment]
Here's my perspective: if a user deletes a Service it means they no longer care/want the corresponding connections to the Service and it's backends anyways (deleting a backend/pod is a different story), in which case I think we should remove the `len(rsList) > 0` check AND the graceful termination check entirely in `cleanupLegacyServices`. Will defer to @lbernail on the final call. Either way moving the `len(rsList) > 0` check seems to fix to the issue. Thanks for working on this @gongguan.

**Comment by gongguan** [comment]
> Here's my perspective: if a user deletes a Service it means they no longer care/want the corresponding connections to the Service and it's backends anyways (deleting a backend/pod is a different story), in which case I think we should remove the `len(rsList) > 0` check AND the graceful termination check entirely in `cleanupLegacyServices`. Will defer to @lbernail on the final call. Either way moving the `len(rsList) > 0` check seems to fix to the issue. Thanks for working on this @gongguan.

Agree

**Comment by gongguan** [comment]
@lbernail to make the scenario clearly, I have create and delete the same service several times. And list ipvs.
image
企业微信截图_a79b8fd2-0704-4a48-af2f-d5f2956eb335
these VirtualServices exist in kube-ipvs0 netlink too.

**Comment by andrewsykim** [comment]
@gongguan let's update the PR as agreed earlier, will reach out to @lbernail for review later

**Comment by gongguan** [comment]
> @gongguan let's update the PR as agreed earlier, will reach out to @lbernail for review later

update the commit with removing graceful deletion?

**Comment by andrewsykim** [comment]
> whether update the commit with removing graceful deletion?  

Yes, let's update the PR to also remove the graceful deletion logic on Service deletion.

**Comment by gongguan** [comment]
> Yes, let's update the PR to also remove the graceful deletion logic on Service deletion.
Done

**Comment by andrewsykim** [comment]
Can you update the release notes please

**Comment by andrewsykim** [comment]
And the commit message too please, something like:
```
In ipvs mode, when service deleted but rs(pod) exist, related ipvs rule and address bound in ipvs0 will not be deleted.
```
Fix a bug in the IPVS proxier where virtual servers are not cleaned up even though the corresponding Service object was deleted.
```

**Comment by gongguan** [comment]
> And the commit message too please, something like:
> 
> ```
> Fix a bug in the IPVS proxier where virtual servers are not cleaned up even though the corresponding Service object was deleted.
> ```

Done

**Comment by gongguan** [comment]
> Can you update the release notes please

Done

**Comment by andrewsykim** [comment]
/approve

Leaving final lgtm to @lbernail

**Comment by lbernail** [comment]
This seems completely reasonable but I think it will introduce a behavior that is different from iptables. In iptables, removing a service will delete the iptables rules but I don't think it will purge the conntrack which will allow existing connections to terminate gracefully. I'm not sure what the expected behavior is.

I also expected a service deletion to trigger a deletion of the matching endpoint resource which should trigger graceful termination. I'm going to investigate why this is not happening. In any case the current behavior is definitely a bug.

Sorry for the delay I haven't had much time lately

**Comment by lbernail** [comment]
OK I confirm that service deletion doesn't trigger a deletion of the real servers which I assumed would have been triggered here: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L768-L770

I'm going to debug this with a custom build

**Comment by lbernail** [comment]
Also, it would be great to add a test case for this so we never have a regression (this is a pretty bad bug...)

**Comment by andrewsykim** [comment]
> I also expected a service deletion to trigger a deletion of the matching endpoint resource which should trigger graceful termination. I

It does, but for TCP it takes 15m for those endpoints to be gracefully terminated and so the virtual server doesn't delete for 15m as well. So I think either we delete the virtual server right away (this PR) or we decrease the graceful termination timeout to a lower value.

**Comment by lbernail** [comment]
@andrewsykim in the test I did, gracefultermination was not triggered at all (and weight remains at 1) so I suspect something is wrong in the handler

Anyway, if we agree we can remove services without gracefultermination this is not important

The PR looks good. @gongguan can you add a test for this specific scenario?

**Comment by lbernail** [comment]
@andrewsykim OK I think I know what happens: `syncProxyRules` is called for service deletion and endpoint deletion, but the reconcile loop only considers active services: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/ipvs/proxier.go#L902

So the only way to trigger gracefulTermination for real servers on service deletion would be in `cleanLegacyService`

**Comment by andrewsykim** [comment]
Thanks for clarifying @lbernail! 

> Anyway, if we agree we can remove services without gracefultermination this is not important

This is my take, if a Service is deleted there's no reason to try graceful termination IMO. I will ask some folks in SIG Network though to confirm if this shouldn't be the case. 

> The PR looks good. @gongguan can you add a test for this specific scenario?

Added a test case for this here https://github.com/kubernetes/kubernetes/pull/81309, PTAL :) 

/lgtm

Thanks @gongguan!

**Comment by lbernail** [comment]
Thanks for adding the test!
We'll need to backport this to all supported branches

**Comment by lbernail** [comment]
I just remember the context behind the graceful termination behavior: https://github.com/kubernetes/kubernetes/issues/71894
TL;DR we did not have `expirenodestconn` before which means when we removed a VS (and therefore RS) on service deletion, established connections were not terminated and their traffic was being blackholed.

Because we now set `expirenodestconn`, established connections will be terminated as soon as a packet is sent

**Comment by gongguan** [comment]
> Added a test case for this here #81309, PTAL :)

Thanks for @andrewsykim 's test. By the way, what can I do for you?

**Comment by andrewsykim** [comment]
Not much from here, PR should be merged by the bot shortly :). Thanks @gongguan!

**Comment by JokerDevops** [comment]
So how do I make sure this problem is solved now? In which version did the solution arrive?

**Comment by JokerDevops** [comment]
[Automatically clears unwanted ipvs rules](https://github.com/kubernetes/kubernetes/issues/89947)

**Comment by Davidrjx** [comment]
part of issues still exists, address bound in kube-ipvs0 will not be deleted in kubernetes1.18.0

**Comment by willzhang** [comment]
SAME IN kubernetes1.23.5

**Comment by andrewsykim** [comment]
Please open a new issue with the details included and assign it to me and @uablrek

**Comment by uablrek** [comment]
I can't reproduce this in v1.23.5. I tried;
* Connect via tcp to a ClusterIP and let the connection stay open
* `watch ip addr show dev kube-ipvs0` in another terminal
* `kubectl delete svc-name`
The ClusterIP is removed from `kube-ipvs0` almost immediately. The tcp connection is black-holed.
